File: python/robot-name/robot_name.py
from random import choice, seed
from string import ascii_uppercase, digits


def generate():
    seed()
    return "".join(
        [
            choice(ascii_uppercase),
            choice(ascii_uppercase),
            choice(digits),
            choice(digits),
            choice(digits),
        ]
    )


class Robot:
    def __init__(self):
        self.name = generate()

    def reset(self):
        self.name = generate()

File: scala/robot-name/src/main/scala/RobotName.scala
import Robot.generate

import scala.collection.mutable
import scala.util.Random

object Robot {
  private val alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  private val numeric = "0123456789"

  private val recipe: Seq[String] = Seq(alpha, alpha, numeric, numeric, numeric)
  private val maxNumberOfNames = recipe.foldLeft(1)(_ * _.length)
  private val generated: mutable.HashSet[String] = mutable.HashSet.empty

  def generate(): String = {
    if (generated.size == maxNumberOfNames) {
      throw new UnsupportedOperationException("Robot names exhausted")
    }

    val name = Stream.continually {
      recipe.map(chars => chars.charAt(Random.nextInt(chars.length))).mkString("")
    }
      .filter { !generated.contains(_) }
      .head

    generated += name
    name
  }
}

class Robot {
  private var _name: String = generate()

  def name: String = _name

  def reset(): Unit = {
    this._name = generate()
  }
}